KAFKA-16339: [2/4 KStream#flatTransform] Remove Deprecated "transformer" methods and classes

fonsdant opened the pull request:
This PR has as base the PR: #17198.

fonsdant committed:
remove flat transform from kstream

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
remove flat transform from kstream impl

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
remove flat transform from scala kstream

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
refactor and remove flat transform from kstream impl test

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
remove flat transform from scala kstream impl test

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
remove flat transform from javadoc of kstream

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant committed:
remove flat transform from javadoc of connected store provider

Signed-off-by: Joao Pedro Fonseca Dantas <fonsdant@gmail.com>

fonsdant commented:
Hi, @mjsax! I have rebased this branch and removed flatTransform and refactored where needed.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java:
```diff
@@ -1576,139 +1562,49 @@ public void shouldProcessFromSourcesThatMatchMultiplePattern() {
-    @SuppressWarnings("deprecation")
-    public void shouldNotAllowBadTransformerSupplierOnFlatTransform() {
-        final org.apache.kafka.streams.kstream.Transformer<String, String, Iterable<KeyValue<String, String>>> transformer = flatTransformerSupplier.get();
+    public void shouldNotAllowBadProcessSupplierOnProcess() {
```
Is there no existing test for `ProcessSupplier` to test test case? Just wondering if this rewrite makes sense (and closes a test gap), or if this rewrite would duplicate an existing test (and thus, we could just remove this test instead of rewriting it)?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java:
```diff
@@ -1576,139 +1562,49 @@ public void shouldProcessFromSourcesThatMatchMultiplePattern() {
-    @SuppressWarnings("deprecation")
-    public void shouldNotAllowBadTransformerSupplierOnFlatTransformWithNamed() {
-        final org.apache.kafka.streams.kstream.Transformer<String, String, Iterable<KeyValue<String, String>>> transformer = flatTransformerSupplier.get();
+    public void shouldNotAllowBadProcessSupplierOnProcessWithNamed() {
```
same question

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java:
```diff
@@ -1576,139 +1562,49 @@ public void shouldProcessFromSourcesThatMatchMultiplePattern() {
-    @SuppressWarnings("deprecation")
-    public void shouldNotAllowBadTransformerSupplierOnFlatTransformWithNamedAndStores() {
-        final org.apache.kafka.streams.kstream.Transformer<String, String, Iterable<KeyValue<String, String>>> transformer = flatTransformerSupplier.get();
+    public void shouldNotAllowBadProcessSupplierOnProcessWithNamedAndStores() {
```
as above

mjsax reviewed (commented):
Thanks for update the PR. Overall LGTM. Just one question about rewriting vs removing some tests.

fonsdant commented:
@mjsax, I have refactored only those which seemed not covered yet. To discover them, I have compared all tests referencing process and flatTransform in KStreamImplTest and removed equivalent ones.

mjsax commented:
Thanks for double checking the tests!

Merged to `trunk`.

fonsdant commented:
Thanks, Matthias!

Two down, two to go 😃
